api/v1beta1/conversion.go

```diff
@@ -16,14 +16,187 @@ limitations under the License.
-func (*CloudStackMachineList) Hub() {}
+	out.Status = v1beta2.CloudStackClusterStatus{
+		FailureDomains: in.Status.FailureDomains,
+		Ready:          in.Status.Ready,
+	}
+	return nil
+}
+
+//nolint:golint,revive,stylecheck
+func Convert_v1beta2_CloudStackCluster_To_v1beta1_CloudStackCluster(in *v1beta2.CloudStackCluster, out *CloudStackCluster, scope conv.Scope) error {
+	out.ObjectMeta = in.ObjectMeta
+	out.Spec = CloudStackClusterSpec{
+		Zones:                getZones(in),
+		ControlPlaneEndpoint: in.Spec.ControlPlaneEndpoint,
+		Account:              in.Spec.Account,
+		Domain:               in.Spec.Domain,
+	}
+
+	out.Status = CloudStackClusterStatus{
+		FailureDomains: in.Status.FailureDomains,
+		Ready:          in.Status.Ready,
+	}
+	return nil
+}
+
+// getZones maps failure domains to zones
+func getZones(csCluster *v1beta2.CloudStackCluster) []Zone {
+	var zones []Zone
+	for _, failureDomain := range csCluster.Spec.FailureDomains {
+		zone := failureDomain.Zone
+		zones = append(zones, Zone{
+			Name: zone.Name,
+			ID:   zone.ID,
+			Network: Network{
+				Name: zone.Network.Name,
+				ID:   zone.Network.ID,
+				Type: zone.Network.Type,
+			},
+		})
+	}
+	return zones
+}
+
+// getFailureDomains maps v1beta1 zones to v1beta2 failure domains.
+func getFailureDomains(csCluster *CloudStackCluster) ([]v1beta2.CloudStackFailureDomainSpec, error) {
+	var failureDomains []v1beta2.CloudStackFailureDomainSpec
+	namespace := csCluster.Namespace
+	for _, zone := range csCluster.Spec.Zones {
+		name, err := GetDefaultFailureDomainName(namespace, csCluster.Name, zone.ID, zone.Name)
+		if err != nil {
+			return nil, err
+		}
+		failureDomains = append(failureDomains, v1beta2.CloudStackFailureDomainSpec{
+			Name: name,
+			Zone: v1beta2.CloudStackZoneSpec{
+				ID:   zone.ID,
+				Name: zone.Name,
+				Network: v1beta2.Network{
+					ID:   zone.Network.ID,
+					Name: zone.Network.Name,
+					Type: zone.Network.Type,
+				},
+			},
+			Domain:  csCluster.Spec.Domain,
+			Account: csCluster.Spec.Account,
+			ACSEndpoint: corev1.SecretReference{
+				Namespace: csCluster.ObjectMeta.Namespace,
+				Name:      DefaultEndpointCredential,
+			},
+		})
+	}
+	return failureDomains, nil
+}
+
+// GetDefaultFailureDomainName return zoneID as failuredomain name.
+// Default failure domain name is used when migrating an old cluster to a multiple-endpoints supported cluster, that
+// requires to convert each zone to a failure domain.
+// When upgrading cluster using eks-a, a secret named global will be created by eks-a, and it is used by following
+// method to get zoneID by calling cloudstack API.
+// When upgrading cluster using clusterctl directly, zoneID is fetched directly from kubernetes cluster in cloudstackzones.
+func GetDefaultFailureDomainName(namespace string, clusterName string, zoneID string, zoneName string) (string, error) {
+	if len(zoneID) > 0 {
+		return zoneID + "-" + clusterName, nil
```

Comment:
Previously it was just zoneId, right? Will this require a corresponding change in https://github.com/maxdrib/eks-anywhere/blob/capc-v1beta2/pkg/api/v1alpha1/cloudstackdatacenterconfig_webhook.go#L91?

Reply:
yes, to distinguish failuredomain with same zone id in different cluster (they are all in eksa-system namespace), a cluster-name is appended. 

The code from your branch still works when upgrading, so I think it is fine.

Reply:
Take last 8 chars of zoneID when creating FD name since there is a name length restriction (<64) in k8s.